{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Setup\" data-toc-modified-id=\"Setup-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Setup</a></span><ul class=\"toc-item\"><li><span><a href=\"#Check-your-sample-table-(i.e.-metadata-file)\" data-toc-modified-id=\"Check-your-sample-table-(i.e.-metadata-file)-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Check your sample table (i.e. metadata file)</a></span></li><li><span><a href=\"#Check-your-TRN\" data-toc-modified-id=\"Check-your-TRN-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Check your TRN</a></span></li><li><span><a href=\"#Load-the-data\" data-toc-modified-id=\"Load-the-data-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Load the data</a></span></li></ul></li><li><span><a href=\"#Regulatory-iModulons\" data-toc-modified-id=\"Regulatory-iModulons-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Regulatory iModulons</a></span></li><li><span><a href=\"#Functional-iModulons\" data-toc-modified-id=\"Functional-iModulons-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Functional iModulons</a></span><ul class=\"toc-item\"><li><span><a href=\"#GO-Enrichments\" data-toc-modified-id=\"GO-Enrichments-3.1\"><span class=\"toc-item-num\">3.1&nbsp;&nbsp;</span>GO Enrichments</a></span></li><li><span><a href=\"#KEGG-Enrichments\" data-toc-modified-id=\"KEGG-Enrichments-3.2\"><span class=\"toc-item-num\">3.2&nbsp;&nbsp;</span>KEGG Enrichments</a></span><ul class=\"toc-item\"><li><span><a href=\"#Load-KEGG-mapping\" data-toc-modified-id=\"Load-KEGG-mapping-3.2.1\"><span class=\"toc-item-num\">3.2.1&nbsp;&nbsp;</span>Load KEGG mapping</a></span></li><li><span><a href=\"#Perform-enrichment\" data-toc-modified-id=\"Perform-enrichment-3.2.2\"><span class=\"toc-item-num\">3.2.2&nbsp;&nbsp;</span>Perform enrichment</a></span></li><li><span><a href=\"#Convert-KEGG-IDs-to-human-readable-names\" data-toc-modified-id=\"Convert-KEGG-IDs-to-human-readable-names-3.2.3\"><span class=\"toc-item-num\">3.2.3&nbsp;&nbsp;</span>Convert KEGG IDs to human-readable names</a></span></li></ul></li><li><span><a href=\"#SubtiWiki-categories\" data-toc-modified-id=\"SubtiWiki-categories-3.3\"><span class=\"toc-item-num\">3.3&nbsp;&nbsp;</span>SubtiWiki categories</a></span></li><li><span><a href=\"#Save-files\" data-toc-modified-id=\"Save-files-3.4\"><span class=\"toc-item-num\">3.4&nbsp;&nbsp;</span>Save files</a></span></li></ul></li><li><span><a href=\"#Check-for-single-gene-iModulons\" data-toc-modified-id=\"Check-for-single-gene-iModulons-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Check for single gene iModulons</a></span></li><li><span><a href=\"#Save-iModulon-object\" data-toc-modified-id=\"Save-iModulon-object-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Save iModulon object</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "This IPython notebook will walk through the steps of characterizing iModulons through the semi-automated tools in PyModulon. You will need:\n",
    "\n",
    "* M and A matrices\n",
    "* Expression data (e.g. `log_tpm_norm.csv`)\n",
    "* Gene table and KEGG/GO annotations (Generated in `gene_annotation.ipynb`)\n",
    "* Sample table, with a column for `project` and `condition`\n",
    "* TRN file\n",
    "\n",
    "Optional:\n",
    "* iModulon table (if you already have some characterized iModulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.core import IcaData\n",
    "from pymodulon.plotting import *\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import re\n",
    "from Bio.KEGG import REST\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the location of your data here\n",
    "data_dir = path.join('..','data','processed_data')\n",
    "\n",
    "# GO and KEGG annotations are in the 'external' folder\n",
    "external_data = path.join('..','data','external')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your sample table (i.e. metadata file)\n",
    "Your metadata file will probably have a lot of columns, most of which you may not care about. Feel free to save a secondary copy of your metadata file with only columns that seem relevant to you. The two most important columns are:\n",
    "1. `project`\n",
    "2. `condition`\n",
    "\n",
    "Make sure that these columns exist in your metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERX1819430</th>\n",
       "      <td>spbeta</td>\n",
       "      <td>spbeta_nopep_20m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERX1819431</th>\n",
       "      <td>spbeta</td>\n",
       "      <td>spbeta_pep_20m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERX1819432</th>\n",
       "      <td>spbeta</td>\n",
       "      <td>spbeta_nopep_20m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERX1819433</th>\n",
       "      <td>spbeta</td>\n",
       "      <td>spbeta_pep_20m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERX1819434</th>\n",
       "      <td>spbeta</td>\n",
       "      <td>phi3T_pep_10m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           project         condition\n",
       "ERX1819430  spbeta  spbeta_nopep_20m\n",
       "ERX1819431  spbeta    spbeta_pep_20m\n",
       "ERX1819432  spbeta  spbeta_nopep_20m\n",
       "ERX1819433  spbeta    spbeta_pep_20m\n",
       "ERX1819434  spbeta     phi3T_pep_10m"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata = pd.read_csv(path.join(data_dir,'metadata.tsv'),index_col=0,sep='\\t')\n",
    "df_metadata[['project','condition']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_metadata.project.notnull().all())\n",
    "print(df_metadata.condition.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your TRN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row of the TRN file represents a regulatory interaction.  \n",
    "**Your TRN file must have the following columns:**\n",
    "1. `regulator` - Name of regulator (`/` or `+` characters will be converted to `;`)\n",
    "1. `gene_id` - Locus tag of gene being regulated\n",
    "\n",
    "The following columns are optional, but are helpful to have:\n",
    "1. `regulator_id` - Locus tag of regulator\n",
    "1. `gene_name` - Name of gene (can automatically update this using `name2num`)\n",
    "1. `direction` - Direction of regulation ('+' for activation, '-' for repression, '?' or NaN for unknown)\n",
    "1. `evidence` - Evidence of regulation (e.g. ChIP-exo, qRT-PCR, SELEX, Motif search)\n",
    "1. `PMID` - Reference for regulation\n",
    "\n",
    "You may add any other columns that could help you. TRNs may be saved as either CSV or TSV files. See below for an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>effect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PerR</td>\n",
       "      <td>BSU_08820</td>\n",
       "      <td>repression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PerR</td>\n",
       "      <td>BSU_32990</td>\n",
       "      <td>repression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PerR</td>\n",
       "      <td>BSU_03510</td>\n",
       "      <td>activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PerR</td>\n",
       "      <td>BSU_11500</td>\n",
       "      <td>repression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PerR</td>\n",
       "      <td>BSU_03520</td>\n",
       "      <td>activation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  regulator    gene_id      effect\n",
       "0      PerR  BSU_08820  repression\n",
       "1      PerR  BSU_32990  repression\n",
       "2      PerR  BSU_03510  activation\n",
       "3      PerR  BSU_11500  repression\n",
       "4      PerR  BSU_03520  activation"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_trn = pd.read_csv(path.join(external_data,'TRN.csv'))\n",
    "df_trn.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `regulator` and `gene_id` must be filled in for each row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_trn.regulator.notnull().all())\n",
    "print(df_trn.gene_id.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data\n",
    "You're now ready to load your IcaData object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Optimizing iModulon thresholds, may take 2-3 minutes...\n"
     ]
    }
   ],
   "source": [
    "ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "                   A = path.join(data_dir,'A.csv'),\n",
    "                   X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "                   gene_table = path.join(data_dir,'gene_info.csv'),\n",
    "                   sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "                   trn = path.join(external_data,'TRN.csv'),\n",
    "                   optimize_cutoff=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you don't have a TRN (or have a very minimal TRN), use `threshold_method = 'kmeans'`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "#                    A = path.join(data_dir,'A.csv'),\n",
    "#                    X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "#                    gene_table = path.join(data_dir,'gene_info.csv'),\n",
    "#                    sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "#                    trn = path.join(data_dir,'TRN.csv'),\n",
    "#                    threshold_method = 'kmeans')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Regulatory iModulons\n",
    "Use `compute_trn_enrichment` to automatically check for Regulatory iModulons. The more complete your TRN, the more regulatory iModulons you'll find."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>S-box</td>\n",
       "      <td>1.436720e-30</td>\n",
       "      <td>8.620321e-30</td>\n",
       "      <td>0.652174</td>\n",
       "      <td>0.576923</td>\n",
       "      <td>0.612245</td>\n",
       "      <td>15.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>ThrR</td>\n",
       "      <td>5.239175e-07</td>\n",
       "      <td>1.571752e-06</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>WalR</td>\n",
       "      <td>1.897147e-12</td>\n",
       "      <td>2.656006e-11</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>7.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>RemA</td>\n",
       "      <td>8.460083e-12</td>\n",
       "      <td>5.922058e-11</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.241379</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>7.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>OpcR</td>\n",
       "      <td>1.456062e-08</td>\n",
       "      <td>6.794954e-08</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>66</td>\n",
       "      <td>AbrB</td>\n",
       "      <td>2.100153e-07</td>\n",
       "      <td>7.875574e-07</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.044280</td>\n",
       "      <td>0.079208</td>\n",
       "      <td>12.0</td>\n",
       "      <td>271.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>66</td>\n",
       "      <td>IolR</td>\n",
       "      <td>1.685319e-06</td>\n",
       "      <td>4.213298e-06</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.177778</td>\n",
       "      <td>4.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>66</td>\n",
       "      <td>LicR</td>\n",
       "      <td>1.465024e-06</td>\n",
       "      <td>4.213298e-06</td>\n",
       "      <td>0.093750</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>67</td>\n",
       "      <td>KipR</td>\n",
       "      <td>6.440803e-21</td>\n",
       "      <td>3.220402e-20</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>67</td>\n",
       "      <td>TnrA</td>\n",
       "      <td>2.543985e-16</td>\n",
       "      <td>6.359963e-16</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>9.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>131 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     imodulon regulator        pvalue        qvalue  precision    recall  \\\n",
       "0           0     S-box  1.436720e-30  8.620321e-30   0.652174  0.576923   \n",
       "1           0      ThrR  5.239175e-07  1.571752e-06   0.130435  0.750000   \n",
       "2           2      WalR  1.897147e-12  2.656006e-11   0.388889  0.291667   \n",
       "3           2      RemA  8.460083e-12  5.922058e-11   0.388889  0.241379   \n",
       "4           2      OpcR  1.456062e-08  6.794954e-08   0.222222  0.500000   \n",
       "..        ...       ...           ...           ...        ...       ...   \n",
       "126        66      AbrB  2.100153e-07  7.875574e-07   0.375000  0.044280   \n",
       "127        66      IolR  1.685319e-06  4.213298e-06   0.125000  0.307692   \n",
       "128        66      LicR  1.465024e-06  4.213298e-06   0.093750  0.750000   \n",
       "129        67      KipR  6.440803e-21  3.220402e-20   0.777778  1.000000   \n",
       "130        67      TnrA  2.543985e-16  6.359963e-16   1.000000  0.107143   \n",
       "\n",
       "      f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0    0.612245  15.0          26.0           23.0     1.0  \n",
       "1    0.222222   3.0           4.0           23.0     1.0  \n",
       "2    0.333333   7.0          24.0           18.0     1.0  \n",
       "3    0.297872   7.0          29.0           18.0     1.0  \n",
       "4    0.307692   4.0           8.0           18.0     1.0  \n",
       "..        ...   ...           ...            ...     ...  \n",
       "126  0.079208  12.0         271.0           32.0     1.0  \n",
       "127  0.177778   4.0          13.0           32.0     1.0  \n",
       "128  0.166667   3.0           4.0           32.0     1.0  \n",
       "129  0.875000   7.0           7.0            9.0     1.0  \n",
       "130  0.193548   9.0          84.0            9.0     1.0  \n",
       "\n",
       "[131 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.compute_trn_enrichment()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also search for AND/OR combinations of regulators using the `max_regs` argument.\n",
    "\n",
    "Regulator enrichments can be directly saved to the `imodulon_table` using the `save` argument. This saves the enrichment with the lowest q-value to the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>S-box</td>\n",
       "      <td>1.436720e-30</td>\n",
       "      <td>8.620321e-30</td>\n",
       "      <td>0.652174</td>\n",
       "      <td>0.576923</td>\n",
       "      <td>0.612245</td>\n",
       "      <td>15.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>ThrR</td>\n",
       "      <td>5.239175e-07</td>\n",
       "      <td>1.571752e-06</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>WalR</td>\n",
       "      <td>1.897147e-12</td>\n",
       "      <td>2.656006e-11</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>7.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>RemA</td>\n",
       "      <td>8.460083e-12</td>\n",
       "      <td>5.922058e-11</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.241379</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>7.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>OpcR</td>\n",
       "      <td>1.456062e-08</td>\n",
       "      <td>6.794954e-08</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>66</td>\n",
       "      <td>AbrB</td>\n",
       "      <td>2.100153e-07</td>\n",
       "      <td>7.875574e-07</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.044280</td>\n",
       "      <td>0.079208</td>\n",
       "      <td>12.0</td>\n",
       "      <td>271.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>66</td>\n",
       "      <td>IolR</td>\n",
       "      <td>1.685319e-06</td>\n",
       "      <td>4.213298e-06</td>\n",
       "      <td>0.125000</td>\n",
       "      <td>0.307692</td>\n",
       "      <td>0.177778</td>\n",
       "      <td>4.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>66</td>\n",
       "      <td>LicR</td>\n",
       "      <td>1.465024e-06</td>\n",
       "      <td>4.213298e-06</td>\n",
       "      <td>0.093750</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>67</td>\n",
       "      <td>KipR</td>\n",
       "      <td>6.440803e-21</td>\n",
       "      <td>3.220402e-20</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>67</td>\n",
       "      <td>TnrA</td>\n",
       "      <td>2.543985e-16</td>\n",
       "      <td>6.359963e-16</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>0.193548</td>\n",
       "      <td>9.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>131 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     imodulon regulator        pvalue        qvalue  precision    recall  \\\n",
       "0           0     S-box  1.436720e-30  8.620321e-30   0.652174  0.576923   \n",
       "1           0      ThrR  5.239175e-07  1.571752e-06   0.130435  0.750000   \n",
       "2           2      WalR  1.897147e-12  2.656006e-11   0.388889  0.291667   \n",
       "3           2      RemA  8.460083e-12  5.922058e-11   0.388889  0.241379   \n",
       "4           2      OpcR  1.456062e-08  6.794954e-08   0.222222  0.500000   \n",
       "..        ...       ...           ...           ...        ...       ...   \n",
       "126        66      AbrB  2.100153e-07  7.875574e-07   0.375000  0.044280   \n",
       "127        66      IolR  1.685319e-06  4.213298e-06   0.125000  0.307692   \n",
       "128        66      LicR  1.465024e-06  4.213298e-06   0.093750  0.750000   \n",
       "129        67      KipR  6.440803e-21  3.220402e-20   0.777778  1.000000   \n",
       "130        67      TnrA  2.543985e-16  6.359963e-16   1.000000  0.107143   \n",
       "\n",
       "      f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0    0.612245  15.0          26.0           23.0     1.0  \n",
       "1    0.222222   3.0           4.0           23.0     1.0  \n",
       "2    0.333333   7.0          24.0           18.0     1.0  \n",
       "3    0.297872   7.0          29.0           18.0     1.0  \n",
       "4    0.307692   4.0           8.0           18.0     1.0  \n",
       "..        ...   ...           ...            ...     ...  \n",
       "126  0.079208  12.0         271.0           32.0     1.0  \n",
       "127  0.177778   4.0          13.0           32.0     1.0  \n",
       "128  0.166667   3.0           4.0           32.0     1.0  \n",
       "129  0.875000   7.0           7.0            9.0     1.0  \n",
       "130  0.193548   9.0          84.0            9.0     1.0  \n",
       "\n",
       "[131 rows x 11 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First search for regulator enrichments with 2 regulators\n",
    "ica_data.compute_trn_enrichment(max_regs=2,save=True)\n",
    "\n",
    "# Next, search for regulator enrichments with just one regulator. This will supercede the 2 regulator enrichments.\n",
    "ica_data.compute_trn_enrichment(max_regs=1,save=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The list of regulatory iModulons are shown below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "72 Total iModulons\n",
      "55 Regulatory iModulons\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-box</td>\n",
       "      <td>1.436720e-30</td>\n",
       "      <td>8.620321e-30</td>\n",
       "      <td>0.652174</td>\n",
       "      <td>0.576923</td>\n",
       "      <td>0.612245</td>\n",
       "      <td>15.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WalR</td>\n",
       "      <td>1.897147e-12</td>\n",
       "      <td>2.656006e-11</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>7.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rex</td>\n",
       "      <td>1.416977e-19</td>\n",
       "      <td>7.084883e-19</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>7.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Fnr</td>\n",
       "      <td>1.066741e-14</td>\n",
       "      <td>6.400444e-14</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>6.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CsoR/Fnr</td>\n",
       "      <td>1.842216e-11</td>\n",
       "      <td>8.799896e-07</td>\n",
       "      <td>0.113636</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.176991</td>\n",
       "      <td>10.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SigM</td>\n",
       "      <td>1.614855e-27</td>\n",
       "      <td>2.099311e-26</td>\n",
       "      <td>0.894737</td>\n",
       "      <td>0.182796</td>\n",
       "      <td>0.303571</td>\n",
       "      <td>17.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EAR riboswitch</td>\n",
       "      <td>3.154667e-40</td>\n",
       "      <td>2.208267e-39</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SigB</td>\n",
       "      <td>4.367180e-66</td>\n",
       "      <td>6.987488e-65</td>\n",
       "      <td>0.962264</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.372263</td>\n",
       "      <td>51.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PyrR</td>\n",
       "      <td>6.910434e-27</td>\n",
       "      <td>1.382087e-26</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.900000</td>\n",
       "      <td>0.947368</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SigV</td>\n",
       "      <td>2.324641e-12</td>\n",
       "      <td>3.022033e-11</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>7.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CodY</td>\n",
       "      <td>2.915811e-73</td>\n",
       "      <td>6.997947e-72</td>\n",
       "      <td>0.821918</td>\n",
       "      <td>0.320856</td>\n",
       "      <td>0.461538</td>\n",
       "      <td>60.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ManR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CymR</td>\n",
       "      <td>9.394653e-42</td>\n",
       "      <td>7.515722e-41</td>\n",
       "      <td>0.758621</td>\n",
       "      <td>0.523810</td>\n",
       "      <td>0.619718</td>\n",
       "      <td>22.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>RhgR</td>\n",
       "      <td>4.739042e-13</td>\n",
       "      <td>1.611274e-11</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.117647</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CymR</td>\n",
       "      <td>1.095370e-35</td>\n",
       "      <td>5.476849e-35</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.571429</td>\n",
       "      <td>18.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>stringent response</td>\n",
       "      <td>1.783190e-45</td>\n",
       "      <td>7.132759e-45</td>\n",
       "      <td>0.914286</td>\n",
       "      <td>0.217687</td>\n",
       "      <td>0.351648</td>\n",
       "      <td>32.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>PhoP</td>\n",
       "      <td>3.398363e-31</td>\n",
       "      <td>1.359345e-30</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.253731</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>17.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SigB</td>\n",
       "      <td>6.792568e-39</td>\n",
       "      <td>1.494365e-37</td>\n",
       "      <td>0.696429</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.281588</td>\n",
       "      <td>39.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>YonO</td>\n",
       "      <td>4.827236e-82</td>\n",
       "      <td>9.654471e-82</td>\n",
       "      <td>0.902439</td>\n",
       "      <td>0.902439</td>\n",
       "      <td>0.902439</td>\n",
       "      <td>37.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TnrA</td>\n",
       "      <td>2.055512e-31</td>\n",
       "      <td>2.466615e-30</td>\n",
       "      <td>0.605263</td>\n",
       "      <td>0.273810</td>\n",
       "      <td>0.377049</td>\n",
       "      <td>23.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>YlxR</td>\n",
       "      <td>4.461785e-16</td>\n",
       "      <td>1.338536e-15</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.347826</td>\n",
       "      <td>8.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>MtlR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ImmR</td>\n",
       "      <td>3.731269e-42</td>\n",
       "      <td>3.731269e-42</td>\n",
       "      <td>0.941176</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>0.914286</td>\n",
       "      <td>16.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SigD</td>\n",
       "      <td>3.053879e-92</td>\n",
       "      <td>6.107758e-91</td>\n",
       "      <td>0.837838</td>\n",
       "      <td>0.516667</td>\n",
       "      <td>0.639175</td>\n",
       "      <td>62.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>TnrA</td>\n",
       "      <td>8.549620e-30</td>\n",
       "      <td>8.549620e-29</td>\n",
       "      <td>0.656250</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.362069</td>\n",
       "      <td>21.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SwrA</td>\n",
       "      <td>7.051312e-60</td>\n",
       "      <td>4.230787e-59</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>0.877193</td>\n",
       "      <td>25.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>YlxR</td>\n",
       "      <td>1.837961e-25</td>\n",
       "      <td>7.351843e-25</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.489796</td>\n",
       "      <td>12.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ComA</td>\n",
       "      <td>1.002354e-09</td>\n",
       "      <td>1.202825e-08</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.240000</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>6.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SigK</td>\n",
       "      <td>3.833827e-110</td>\n",
       "      <td>6.134123e-109</td>\n",
       "      <td>0.646617</td>\n",
       "      <td>0.646617</td>\n",
       "      <td>0.646617</td>\n",
       "      <td>86.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SigI</td>\n",
       "      <td>1.028044e-11</td>\n",
       "      <td>7.196308e-11</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.294118</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>5.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Zur</td>\n",
       "      <td>5.165927e-33</td>\n",
       "      <td>2.066371e-32</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Thi-box</td>\n",
       "      <td>2.378535e-31</td>\n",
       "      <td>2.378535e-30</td>\n",
       "      <td>0.590909</td>\n",
       "      <td>0.928571</td>\n",
       "      <td>0.722222</td>\n",
       "      <td>13.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NusA</td>\n",
       "      <td>1.788207e-11</td>\n",
       "      <td>7.510469e-10</td>\n",
       "      <td>0.054264</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.102941</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CsoR</td>\n",
       "      <td>1.163609e-10</td>\n",
       "      <td>8.145265e-10</td>\n",
       "      <td>0.127273</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.202899</td>\n",
       "      <td>7.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>SigB</td>\n",
       "      <td>1.100046e-65</td>\n",
       "      <td>3.080128e-64</td>\n",
       "      <td>0.746988</td>\n",
       "      <td>0.280543</td>\n",
       "      <td>0.407895</td>\n",
       "      <td>62.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>FruR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>AcoR</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>LnrK</td>\n",
       "      <td>1.632739e-08</td>\n",
       "      <td>8.163697e-08</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>DegU</td>\n",
       "      <td>2.854706e-11</td>\n",
       "      <td>1.427353e-10</td>\n",
       "      <td>0.370370</td>\n",
       "      <td>0.119048</td>\n",
       "      <td>0.180180</td>\n",
       "      <td>10.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Xpf</td>\n",
       "      <td>1.548917e-55</td>\n",
       "      <td>1.239133e-54</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>27.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Fur</td>\n",
       "      <td>1.213043e-74</td>\n",
       "      <td>1.334347e-73</td>\n",
       "      <td>0.950000</td>\n",
       "      <td>0.622951</td>\n",
       "      <td>0.752475</td>\n",
       "      <td>38.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>CcpA</td>\n",
       "      <td>7.893839e-56</td>\n",
       "      <td>2.683905e-54</td>\n",
       "      <td>0.719512</td>\n",
       "      <td>0.225191</td>\n",
       "      <td>0.343023</td>\n",
       "      <td>59.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CcpN</td>\n",
       "      <td>1.544295e-11</td>\n",
       "      <td>7.721475e-11</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>G-box</td>\n",
       "      <td>6.185621e-42</td>\n",
       "      <td>1.855686e-41</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.937500</td>\n",
       "      <td>0.967742</td>\n",
       "      <td>15.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>ResD</td>\n",
       "      <td>1.151685e-42</td>\n",
       "      <td>2.994381e-41</td>\n",
       "      <td>0.482759</td>\n",
       "      <td>0.560000</td>\n",
       "      <td>0.518519</td>\n",
       "      <td>28.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>SigW</td>\n",
       "      <td>8.102748e-84</td>\n",
       "      <td>7.292473e-83</td>\n",
       "      <td>0.872727</td>\n",
       "      <td>0.607595</td>\n",
       "      <td>0.716418</td>\n",
       "      <td>48.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>SigE</td>\n",
       "      <td>7.457682e-55</td>\n",
       "      <td>1.789844e-53</td>\n",
       "      <td>0.628866</td>\n",
       "      <td>0.255230</td>\n",
       "      <td>0.363095</td>\n",
       "      <td>61.0</td>\n",
       "      <td>239.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Rok</td>\n",
       "      <td>6.692924e-17</td>\n",
       "      <td>1.673231e-15</td>\n",
       "      <td>0.156250</td>\n",
       "      <td>0.441176</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>15.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>GlpP</td>\n",
       "      <td>1.022901e-13</td>\n",
       "      <td>1.022901e-12</td>\n",
       "      <td>0.384615</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.555556</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>YonO</td>\n",
       "      <td>4.492705e-08</td>\n",
       "      <td>8.985409e-08</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.188679</td>\n",
       "      <td>5.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>MntR</td>\n",
       "      <td>1.669070e-15</td>\n",
       "      <td>1.669070e-15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.833333</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>FapR</td>\n",
       "      <td>4.919296e-22</td>\n",
       "      <td>2.951578e-21</td>\n",
       "      <td>0.888889</td>\n",
       "      <td>0.727273</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>8.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>AbrB</td>\n",
       "      <td>6.614330e-22</td>\n",
       "      <td>8.598629e-21</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.073801</td>\n",
       "      <td>0.136054</td>\n",
       "      <td>20.0</td>\n",
       "      <td>271.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>CcpA</td>\n",
       "      <td>4.279881e-27</td>\n",
       "      <td>6.419821e-26</td>\n",
       "      <td>0.812500</td>\n",
       "      <td>0.099237</td>\n",
       "      <td>0.176871</td>\n",
       "      <td>26.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>KipR</td>\n",
       "      <td>6.440803e-21</td>\n",
       "      <td>3.220402e-20</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.875000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             regulator         pvalue         qvalue  precision    recall  \\\n",
       "0                S-box   1.436720e-30   8.620321e-30   0.652174  0.576923   \n",
       "2                 WalR   1.897147e-12   2.656006e-11   0.388889  0.291667   \n",
       "4                  Rex   1.416977e-19   7.084883e-19   1.000000  0.583333   \n",
       "5                  Fnr   1.066741e-14   6.400444e-14   0.600000  0.545455   \n",
       "8             CsoR/Fnr   1.842216e-11   8.799896e-07   0.113636  0.400000   \n",
       "10                SigM   1.614855e-27   2.099311e-26   0.894737  0.182796   \n",
       "11      EAR riboswitch   3.154667e-40   2.208267e-39   0.833333  1.000000   \n",
       "12                SigB   4.367180e-66   6.987488e-65   0.962264  0.230769   \n",
       "13                PyrR   6.910434e-27   1.382087e-26   1.000000  0.900000   \n",
       "14                SigV   2.324641e-12   3.022033e-11   0.333333  0.333333   \n",
       "15                CodY   2.915811e-73   6.997947e-72   0.821918  0.320856   \n",
       "16                ManR   0.000000e+00   0.000000e+00   1.000000  1.000000   \n",
       "18                CymR   9.394653e-42   7.515722e-41   0.758621  0.523810   \n",
       "19                RhgR   4.739042e-13   1.611274e-11   0.062500  1.000000   \n",
       "21                CymR   1.095370e-35   5.476849e-35   0.857143  0.428571   \n",
       "22  stringent response   1.783190e-45   7.132759e-45   0.914286  0.217687   \n",
       "24                PhoP   3.398363e-31   1.359345e-30   0.944444  0.253731   \n",
       "25                SigB   6.792568e-39   1.494365e-37   0.696429  0.176471   \n",
       "26                YonO   4.827236e-82   9.654471e-82   0.902439  0.902439   \n",
       "27                TnrA   2.055512e-31   2.466615e-30   0.605263  0.273810   \n",
       "28                YlxR   4.461785e-16   1.338536e-15   0.800000  0.222222   \n",
       "29                MtlR   0.000000e+00   0.000000e+00   1.000000  1.000000   \n",
       "30                ImmR   3.731269e-42   3.731269e-42   0.941176  0.888889   \n",
       "31                SigD   3.053879e-92   6.107758e-91   0.837838  0.516667   \n",
       "32                TnrA   8.549620e-30   8.549620e-29   0.656250  0.250000   \n",
       "33                SwrA   7.051312e-60   4.230787e-59   1.000000  0.781250   \n",
       "34                YlxR   1.837961e-25   7.351843e-25   0.923077  0.333333   \n",
       "35                ComA   1.002354e-09   1.202825e-08   0.285714  0.240000   \n",
       "36                SigK  3.833827e-110  6.134123e-109   0.646617  0.646617   \n",
       "37                SigI   1.028044e-11   7.196308e-11   0.714286  0.294118   \n",
       "38                 Zur   5.165927e-33   2.066371e-32   0.800000  1.000000   \n",
       "39             Thi-box   2.378535e-31   2.378535e-30   0.590909  0.928571   \n",
       "40                NusA   1.788207e-11   7.510469e-10   0.054264  1.000000   \n",
       "41                CsoR   1.163609e-10   8.145265e-10   0.127273  0.500000   \n",
       "42                SigB   1.100046e-65   3.080128e-64   0.746988  0.280543   \n",
       "43                FruR   0.000000e+00   0.000000e+00   1.000000  1.000000   \n",
       "44                AcoR   0.000000e+00   0.000000e+00   1.000000  1.000000   \n",
       "45                LnrK   1.632739e-08   8.163697e-08   0.250000  1.000000   \n",
       "46                DegU   2.854706e-11   1.427353e-10   0.370370  0.119048   \n",
       "47                 Xpf   1.548917e-55   1.239133e-54   0.500000  1.000000   \n",
       "48                 Fur   1.213043e-74   1.334347e-73   0.950000  0.622951   \n",
       "49                CcpA   7.893839e-56   2.683905e-54   0.719512  0.225191   \n",
       "50                CcpN   1.544295e-11   7.721475e-11   0.666667  0.666667   \n",
       "54               G-box   6.185621e-42   1.855686e-41   1.000000  0.937500   \n",
       "55                ResD   1.151685e-42   2.994381e-41   0.482759  0.560000   \n",
       "56                SigW   8.102748e-84   7.292473e-83   0.872727  0.607595   \n",
       "57                SigE   7.457682e-55   1.789844e-53   0.628866  0.255230   \n",
       "58                 Rok   6.692924e-17   1.673231e-15   0.156250  0.441176   \n",
       "59                GlpP   1.022901e-13   1.022901e-12   0.384615  1.000000   \n",
       "60                YonO   4.492705e-08   8.985409e-08   0.416667  0.121951   \n",
       "63                MntR   1.669070e-15   1.669070e-15   1.000000  0.714286   \n",
       "64                FapR   4.919296e-22   2.951578e-21   0.888889  0.727273   \n",
       "65                AbrB   6.614330e-22   8.598629e-21   0.869565  0.073801   \n",
       "66                CcpA   4.279881e-27   6.419821e-26   0.812500  0.099237   \n",
       "67                KipR   6.440803e-21   3.220402e-20   0.777778  1.000000   \n",
       "\n",
       "     f1score    TP  regulon_size  imodulon_size  n_regs  \n",
       "0   0.612245  15.0          26.0           23.0     1.0  \n",
       "2   0.333333   7.0          24.0           18.0     1.0  \n",
       "4   0.736842   7.0          12.0            7.0     1.0  \n",
       "5   0.571429   6.0          11.0           10.0     1.0  \n",
       "8   0.176991  10.0          25.0           88.0     2.0  \n",
       "10  0.303571  17.0          93.0           19.0     1.0  \n",
       "11  0.909091  15.0          15.0           18.0     1.0  \n",
       "12  0.372263  51.0         221.0           53.0     1.0  \n",
       "13  0.947368   9.0          10.0            9.0     1.0  \n",
       "14  0.333333   7.0          21.0           21.0     1.0  \n",
       "15  0.461538  60.0         187.0           73.0     1.0  \n",
       "16  1.000000   4.0           4.0            4.0     1.0  \n",
       "18  0.619718  22.0          42.0           29.0     1.0  \n",
       "19  0.117647   8.0           8.0          128.0     1.0  \n",
       "21  0.571429  18.0          42.0           21.0     1.0  \n",
       "22  0.351648  32.0         147.0           35.0     1.0  \n",
       "24  0.400000  17.0          67.0           18.0     1.0  \n",
       "25  0.281588  39.0         221.0           56.0     1.0  \n",
       "26  0.902439  37.0          41.0           41.0     1.0  \n",
       "27  0.377049  23.0          84.0           38.0     1.0  \n",
       "28  0.347826   8.0          36.0           10.0     1.0  \n",
       "29  1.000000   3.0           3.0            3.0     1.0  \n",
       "30  0.914286  16.0          18.0           17.0     1.0  \n",
       "31  0.639175  62.0         120.0           74.0     1.0  \n",
       "32  0.362069  21.0          84.0           32.0     1.0  \n",
       "33  0.877193  25.0          32.0           25.0     1.0  \n",
       "34  0.489796  12.0          36.0           13.0     1.0  \n",
       "35  0.260870   6.0          25.0           21.0     1.0  \n",
       "36  0.646617  86.0         133.0          133.0     1.0  \n",
       "37  0.416667   5.0          17.0            7.0     1.0  \n",
       "38  0.888889  12.0          12.0           15.0     1.0  \n",
       "39  0.722222  13.0          14.0           22.0     1.0  \n",
       "40  0.102941   7.0           7.0          129.0     1.0  \n",
       "41  0.202899   7.0          14.0           55.0     1.0  \n",
       "42  0.407895  62.0         221.0           83.0     1.0  \n",
       "43  1.000000   3.0           3.0            3.0     1.0  \n",
       "44  1.000000   4.0           4.0            4.0     1.0  \n",
       "45  0.400000   3.0           3.0           12.0     1.0  \n",
       "46  0.180180  10.0          84.0           27.0     1.0  \n",
       "47  0.666667  27.0          27.0           54.0     1.0  \n",
       "48  0.752475  38.0          61.0           40.0     1.0  \n",
       "49  0.343023  59.0         262.0           82.0     1.0  \n",
       "50  0.666667   4.0           6.0            6.0     1.0  \n",
       "54  0.967742  15.0          16.0           15.0     1.0  \n",
       "55  0.518519  28.0          50.0           58.0     1.0  \n",
       "56  0.716418  48.0          79.0           55.0     1.0  \n",
       "57  0.363095  61.0         239.0           97.0     1.0  \n",
       "58  0.230769  15.0          34.0           96.0     1.0  \n",
       "59  0.555556   5.0           5.0           13.0     1.0  \n",
       "60  0.188679   5.0          41.0           12.0     1.0  \n",
       "63  0.833333   5.0           7.0            5.0     1.0  \n",
       "64  0.800000   8.0          11.0            9.0     1.0  \n",
       "65  0.136054  20.0         271.0           23.0     1.0  \n",
       "66  0.176871  26.0         262.0           32.0     1.0  \n",
       "67  0.875000   7.0           7.0            9.0     1.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]\n",
    "print(len(ica_data.imodulon_table),'Total iModulons')\n",
    "print(len(regulatory_imodulons),'Regulatory iModulons')\n",
    "regulatory_imodulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can rename iModulons in this jupyter notebook, or you can save the iModulon table as a CSV and edit it in Excel.\n",
    "\n",
    "If two iModulons have the same regulator (e.g. 'Reg'), they will be named 'Reg-1' and 'Reg-2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Duplicate iModulon names detected. iModulon 12 will be renamed to SigB-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 18 will be renamed to CymR-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 21 will be renamed to CymR-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 25 will be renamed to SigB-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 26 will be renamed to YonO-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 27 will be renamed to TnrA-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 28 will be renamed to YlxR-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 32 will be renamed to TnrA-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 34 will be renamed to YlxR-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 42 will be renamed to SigB-3\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 49 will be renamed to CcpA-1\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 60 will be renamed to YonO-2\n",
      "WARNING:root:Duplicate iModulon names detected. iModulon 66 will be renamed to CcpA-2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S-box</th>\n",
       "      <td>S-box</td>\n",
       "      <td>1.436720e-30</td>\n",
       "      <td>8.620321e-30</td>\n",
       "      <td>0.652174</td>\n",
       "      <td>0.576923</td>\n",
       "      <td>0.612245</td>\n",
       "      <td>15.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WalR</th>\n",
       "      <td>WalR</td>\n",
       "      <td>1.897147e-12</td>\n",
       "      <td>2.656006e-11</td>\n",
       "      <td>0.388889</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>7.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rex</th>\n",
       "      <td>Rex</td>\n",
       "      <td>1.416977e-19</td>\n",
       "      <td>7.084883e-19</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.736842</td>\n",
       "      <td>7.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      regulator        pvalue        qvalue  precision    recall   f1score  \\\n",
       "S-box     S-box  1.436720e-30  8.620321e-30   0.652174  0.576923  0.612245   \n",
       "1           NaN           NaN           NaN        NaN       NaN       NaN   \n",
       "WalR       WalR  1.897147e-12  2.656006e-11   0.388889  0.291667  0.333333   \n",
       "3           NaN           NaN           NaN        NaN       NaN       NaN   \n",
       "Rex         Rex  1.416977e-19  7.084883e-19   1.000000  0.583333  0.736842   \n",
       "\n",
       "         TP  regulon_size  imodulon_size  n_regs  \n",
       "S-box  15.0          26.0           23.0     1.0  \n",
       "1       NaN           NaN            NaN     NaN  \n",
       "WalR    7.0          24.0           18.0     1.0  \n",
       "3       NaN           NaN            NaN     NaN  \n",
       "Rex     7.0          12.0            7.0     1.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.rename_imodulons(regulatory_imodulons.regulator.to_dict())\n",
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "regulatory_imodulons = ica_data.imodulon_table[ica_data.imodulon_table.regulator.notnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GO annotations and KEGG pathways/modules were generated in the 1_create_the_gene_table.ipynb notebook. Enrichments will be calculated in this notebook, and further curated in the 3_manual_iModulon_curation notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First load the Gene Ontology annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_ontology</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BSU_25810</td>\n",
       "      <td>arsR</td>\n",
       "      <td>regulation of DNA-templated transcription</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BSU_25790</td>\n",
       "      <td>arsB</td>\n",
       "      <td>arsenite transmembrane transporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BSU_25790</td>\n",
       "      <td>arsB</td>\n",
       "      <td>antimonite transmembrane transporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BSU_25790</td>\n",
       "      <td>arsB</td>\n",
       "      <td>antiporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BSU_25790</td>\n",
       "      <td>arsB</td>\n",
       "      <td>plasma membrane</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id gene_name                                  gene_ontology\n",
       "0  BSU_25810      arsR      regulation of DNA-templated transcription\n",
       "1  BSU_25790      arsB    arsenite transmembrane transporter activity\n",
       "2  BSU_25790      arsB  antimonite transmembrane transporter activity\n",
       "3  BSU_25790      arsB                            antiporter activity\n",
       "4  BSU_25790      arsB                                plasma membrane"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO = pd.read_csv(path.join(external_data,'GO_annotations_curated.csv'),index_col=0)\n",
    "DF_GO.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich = ica_data.compute_annotation_enrichment(DF_GO,'gene_ontology')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>gene_ontology</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-box</td>\n",
       "      <td>L-methionine salvage from methylthioadenosine</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.017114</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.160000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-box</td>\n",
       "      <td>carbon-sulfur lyase activity</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.017114</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.160000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-box</td>\n",
       "      <td>transsulfuration</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.034116</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-box</td>\n",
       "      <td>cytoplasm</td>\n",
       "      <td>0.000315</td>\n",
       "      <td>0.099678</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>0.042373</td>\n",
       "      <td>0.070922</td>\n",
       "      <td>5.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rex</td>\n",
       "      <td>cytochrome complex</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.008516</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon                                  gene_ontology    pvalue    qvalue  \\\n",
       "0    S-box  L-methionine salvage from methylthioadenosine  0.000027  0.017114   \n",
       "1    S-box                   carbon-sulfur lyase activity  0.000027  0.017114   \n",
       "2    S-box                               transsulfuration  0.000081  0.034116   \n",
       "3    S-box                                      cytoplasm  0.000315  0.099678   \n",
       "4      Rex                             cytochrome complex  0.000007  0.008516   \n",
       "\n",
       "   precision    recall   f1score   TP  target_set_size  imodulon_size  \n",
       "0   0.086957  1.000000  0.160000  2.0              2.0           23.0  \n",
       "1   0.086957  1.000000  0.160000  2.0              2.0           23.0  \n",
       "2   0.086957  0.666667  0.153846  2.0              3.0           23.0  \n",
       "3   0.217391  0.042373  0.070922  5.0            118.0           23.0  \n",
       "4   0.285714  0.666667  0.400000  2.0              3.0            7.0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KEGG Enrichments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load KEGG mapping\n",
    "The `kegg_mapping.csv` file contains KEGG orthologies, pathways, modules, and reactions. Only pathways and modules are relevant to iModulon characterization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KEGG_pathway' 'KEGG_module' 'KEGG_reaction']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database</th>\n",
       "      <th>kegg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BSU_00010</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map02020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BSU_00010</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map04112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BSU_00020</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BSU_00020</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BSU_00020</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map01100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id      database   kegg_id\n",
       "2   BSU_00010  KEGG_pathway  map02020\n",
       "3   BSU_00010  KEGG_pathway  map04112\n",
       "10  BSU_00020  KEGG_pathway  map00230\n",
       "11  BSU_00020  KEGG_pathway  map00240\n",
       "12  BSU_00020  KEGG_pathway  map01100"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_KEGG = pd.read_csv(path.join(external_data,'kegg_mapping.csv'),index_col=0)\n",
    "print(DF_KEGG.database.unique())\n",
    "DF_KEGG.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways = DF_KEGG[DF_KEGG.database == 'KEGG_pathway']\n",
    "kegg_modules = DF_KEGG[DF_KEGG.database == 'KEGG_module']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform enrichment\n",
    "Uses the `compute_annotation_enrichment` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_pathway_enrich = ica_data.compute_annotation_enrichment(kegg_pathways,'kegg_id')\n",
    "DF_module_enrich = ica_data.compute_annotation_enrichment(kegg_modules,'kegg_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map00270</td>\n",
       "      <td>1.439201e-21</td>\n",
       "      <td>3.641178e-19</td>\n",
       "      <td>0.608696</td>\n",
       "      <td>0.233333</td>\n",
       "      <td>0.337349</td>\n",
       "      <td>14.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map01230</td>\n",
       "      <td>2.062496e-12</td>\n",
       "      <td>2.609057e-10</td>\n",
       "      <td>0.521739</td>\n",
       "      <td>0.079470</td>\n",
       "      <td>0.137931</td>\n",
       "      <td>12.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map01100</td>\n",
       "      <td>3.526404e-11</td>\n",
       "      <td>2.973934e-09</td>\n",
       "      <td>0.826087</td>\n",
       "      <td>0.023960</td>\n",
       "      <td>0.046569</td>\n",
       "      <td>19.0</td>\n",
       "      <td>793.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map00450</td>\n",
       "      <td>2.893362e-10</td>\n",
       "      <td>1.830051e-08</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map00260</td>\n",
       "      <td>1.758714e-09</td>\n",
       "      <td>8.899095e-08</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>0.155556</td>\n",
       "      <td>0.205882</td>\n",
       "      <td>7.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0    S-box  map00270  1.439201e-21  3.641178e-19   0.608696  0.233333   \n",
       "1    S-box  map01230  2.062496e-12  2.609057e-10   0.521739  0.079470   \n",
       "2    S-box  map01100  3.526404e-11  2.973934e-09   0.826087  0.023960   \n",
       "3    S-box  map00450  2.893362e-10  1.830051e-08   0.260870  0.315789   \n",
       "4    S-box  map00260  1.758714e-09  8.899095e-08   0.304348  0.155556   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \n",
       "0  0.337349  14.0             60.0           23.0  \n",
       "1  0.137931  12.0            151.0           23.0  \n",
       "2  0.046569  19.0            793.0           23.0  \n",
       "3  0.285714   6.0             19.0           23.0  \n",
       "4  0.205882   7.0             45.0           23.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00017</td>\n",
       "      <td>2.750054e-13</td>\n",
       "      <td>9.790193e-11</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>0.368421</td>\n",
       "      <td>7.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00034</td>\n",
       "      <td>8.257398e-13</td>\n",
       "      <td>1.469817e-10</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00018</td>\n",
       "      <td>1.081269e-05</td>\n",
       "      <td>1.283106e-03</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00020</td>\n",
       "      <td>2.679498e-04</td>\n",
       "      <td>2.384753e-02</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00609</td>\n",
       "      <td>7.430007e-04</td>\n",
       "      <td>5.290165e-02</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon kegg_id        pvalue        qvalue  precision    recall   f1score  \\\n",
       "0    S-box  M00017  2.750054e-13  9.790193e-11   0.304348  0.466667  0.368421   \n",
       "1    S-box  M00034  8.257398e-13  1.469817e-10   0.304348  0.411765  0.350000   \n",
       "2    S-box  M00018  1.081269e-05  1.283106e-03   0.130435  0.333333  0.187500   \n",
       "3    S-box  M00020  2.679498e-04  2.384753e-02   0.086957  0.400000  0.142857   \n",
       "4    S-box  M00609  7.430007e-04  5.290165e-02   0.086957  0.250000  0.129032   \n",
       "\n",
       "    TP  target_set_size  imodulon_size  \n",
       "0  7.0             15.0           23.0  \n",
       "1  7.0             17.0           23.0  \n",
       "2  3.0              9.0           23.0  \n",
       "3  2.0              5.0           23.0  \n",
       "4  2.0              8.0           23.0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert KEGG IDs to human-readable names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "74f7e181ede44a1490d900d9b3f269c4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/137 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b31fc90567d54736b28e906fd7aae397",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/136 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for idx,key in tqdm(DF_pathway_enrich.kegg_id.items(),total=len(DF_pathway_enrich)):\n",
    "    if '-' not in key:\n",
    "        text = REST.kegg_find('pathway',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = None\n",
    "    \n",
    "for idx,key in tqdm(DF_module_enrich.kegg_id.items(),total=len(DF_module_enrich)):\n",
    "    if '-' not in key:\n",
    "        text = REST.kegg_find('module',key).read()\n",
    "    try:\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_module_enrich.loc[idx,'module_name'] = name\n",
    "    except AttributeError:\n",
    "        DF_module_enrich.loc[idx,'module_name'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map00270</td>\n",
       "      <td>1.439201e-21</td>\n",
       "      <td>3.641178e-19</td>\n",
       "      <td>0.608696</td>\n",
       "      <td>0.233333</td>\n",
       "      <td>0.337349</td>\n",
       "      <td>14.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map01230</td>\n",
       "      <td>2.062496e-12</td>\n",
       "      <td>2.609057e-10</td>\n",
       "      <td>0.521739</td>\n",
       "      <td>0.079470</td>\n",
       "      <td>0.137931</td>\n",
       "      <td>12.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map01100</td>\n",
       "      <td>3.526404e-11</td>\n",
       "      <td>2.973934e-09</td>\n",
       "      <td>0.826087</td>\n",
       "      <td>0.023960</td>\n",
       "      <td>0.046569</td>\n",
       "      <td>19.0</td>\n",
       "      <td>793.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Metabolic pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map00450</td>\n",
       "      <td>2.893362e-10</td>\n",
       "      <td>1.830051e-08</td>\n",
       "      <td>0.260870</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>0.285714</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Selenocompound metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S-box</td>\n",
       "      <td>map00260</td>\n",
       "      <td>1.758714e-09</td>\n",
       "      <td>8.899095e-08</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>0.155556</td>\n",
       "      <td>0.205882</td>\n",
       "      <td>7.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0    S-box  map00270  1.439201e-21  3.641178e-19   0.608696  0.233333   \n",
       "1    S-box  map01230  2.062496e-12  2.609057e-10   0.521739  0.079470   \n",
       "2    S-box  map01100  3.526404e-11  2.973934e-09   0.826087  0.023960   \n",
       "3    S-box  map00450  2.893362e-10  1.830051e-08   0.260870  0.315789   \n",
       "4    S-box  map00260  1.758714e-09  8.899095e-08   0.304348  0.155556   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \\\n",
       "0  0.337349  14.0             60.0           23.0   \n",
       "1  0.137931  12.0            151.0           23.0   \n",
       "2  0.046569  19.0            793.0           23.0   \n",
       "3  0.285714   6.0             19.0           23.0   \n",
       "4  0.205882   7.0             45.0           23.0   \n",
       "\n",
       "                               pathway_name  \n",
       "0        Cysteine and methionine metabolism  \n",
       "1               Biosynthesis of amino acids  \n",
       "2                        Metabolic pathways  \n",
       "3                 Selenocompound metabolism  \n",
       "4  Glycine, serine and threonine metabolism  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00017</td>\n",
       "      <td>2.750054e-13</td>\n",
       "      <td>9.790193e-11</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>0.368421</td>\n",
       "      <td>7.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Methionine biosynthesis, aspartate =&gt; homoseri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00034</td>\n",
       "      <td>8.257398e-13</td>\n",
       "      <td>1.469817e-10</td>\n",
       "      <td>0.304348</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Methionine salvage pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00018</td>\n",
       "      <td>1.081269e-05</td>\n",
       "      <td>1.283106e-03</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Threonine biosynthesis, aspartate =&gt; homoserin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00020</td>\n",
       "      <td>2.679498e-04</td>\n",
       "      <td>2.384753e-02</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Serine biosynthesis, glycerate-3P =&gt; serine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S-box</td>\n",
       "      <td>M00609</td>\n",
       "      <td>7.430007e-04</td>\n",
       "      <td>5.290165e-02</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Cysteine biosynthesis, methionine =&gt; cysteine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  imodulon kegg_id        pvalue        qvalue  precision    recall   f1score  \\\n",
       "0    S-box  M00017  2.750054e-13  9.790193e-11   0.304348  0.466667  0.368421   \n",
       "1    S-box  M00034  8.257398e-13  1.469817e-10   0.304348  0.411765  0.350000   \n",
       "2    S-box  M00018  1.081269e-05  1.283106e-03   0.130435  0.333333  0.187500   \n",
       "3    S-box  M00020  2.679498e-04  2.384753e-02   0.086957  0.400000  0.142857   \n",
       "4    S-box  M00609  7.430007e-04  5.290165e-02   0.086957  0.250000  0.129032   \n",
       "\n",
       "    TP  target_set_size  imodulon_size  \\\n",
       "0  7.0             15.0           23.0   \n",
       "1  7.0             17.0           23.0   \n",
       "2  3.0              9.0           23.0   \n",
       "3  2.0              5.0           23.0   \n",
       "4  2.0              8.0           23.0   \n",
       "\n",
       "                                         module_name  \n",
       "0  Methionine biosynthesis, aspartate => homoseri...  \n",
       "1                         Methionine salvage pathway  \n",
       "2  Threonine biosynthesis, aspartate => homoserin...  \n",
       "3        Serine biosynthesis, glycerate-3P => serine  \n",
       "4      Cysteine biosynthesis, methionine => cysteine  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SubtiWiki categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubtiWiki</th>\n",
       "      <th>BSU_number</th>\n",
       "      <th>FuncId</th>\n",
       "      <th>FuncName1</th>\n",
       "      <th>FuncName2</th>\n",
       "      <th>FuncName3</th>\n",
       "      <th>FuncName4</th>\n",
       "      <th>FuncName5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dat</td>\n",
       "      <td>BSU_09670</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ddl</td>\n",
       "      <td>BSU_04560</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>glmM</td>\n",
       "      <td>BSU_01770</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>glmS</td>\n",
       "      <td>BSU_01780</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mraY</td>\n",
       "      <td>BSU_15190</td>\n",
       "      <td>SW 1.1</td>\n",
       "      <td>Cellular processes</td>\n",
       "      <td>cell envelope and cell division</td>\n",
       "      <td>cell wall synthesis</td>\n",
       "      <td>biosynthesis of peptidoglycan</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SubtiWiki BSU_number  FuncId           FuncName1  \\\n",
       "0       dat  BSU_09670  SW 1.1  Cellular processes   \n",
       "1       ddl  BSU_04560  SW 1.1  Cellular processes   \n",
       "2      glmM  BSU_01770  SW 1.1  Cellular processes   \n",
       "3      glmS  BSU_01780  SW 1.1  Cellular processes   \n",
       "4      mraY  BSU_15190  SW 1.1  Cellular processes   \n",
       "\n",
       "                         FuncName2            FuncName3  \\\n",
       "0  cell envelope and cell division  cell wall synthesis   \n",
       "1  cell envelope and cell division  cell wall synthesis   \n",
       "2  cell envelope and cell division  cell wall synthesis   \n",
       "3  cell envelope and cell division  cell wall synthesis   \n",
       "4  cell envelope and cell division  cell wall synthesis   \n",
       "\n",
       "                       FuncName4 FuncName5  \n",
       "0  biosynthesis of peptidoglycan       NaN  \n",
       "1  biosynthesis of peptidoglycan       NaN  \n",
       "2  biosynthesis of peptidoglycan       NaN  \n",
       "3  biosynthesis of peptidoglycan       NaN  \n",
       "4  biosynthesis of peptidoglycan       NaN  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_subtiwiki = pd.read_csv(path.join(external_data,'subtiwiki_categories.csv'))\n",
    "DF_subtiwiki.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BSU_09670</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BSU_04560</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BSU_01770</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BSU_01780</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BSU_15190</td>\n",
       "      <td>FuncName1</td>\n",
       "      <td>Cellular processes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_id   variable               value\n",
       "0  BSU_09670  FuncName1  Cellular processes\n",
       "1  BSU_04560  FuncName1  Cellular processes\n",
       "2  BSU_01770  FuncName1  Cellular processes\n",
       "3  BSU_01780  FuncName1  Cellular processes\n",
       "4  BSU_15190  FuncName1  Cellular processes"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Change the subtiwiki annotation format into a list of genes and categories\n",
    "DF_subtiwiki = DF_subtiwiki.rename({'BSU_number':'gene_id'},axis=1)\n",
    "DF_subtiwiki = DF_subtiwiki.melt(id_vars='gene_id',value_vars=['FuncName1','FuncName2','FuncName3','FuncName4','FuncName5'])\n",
    "DF_subtiwiki = DF_subtiwiki[DF_subtiwiki.value.notnull() & DF_subtiwiki.gene_id.isin(ica_data.gene_names)]\n",
    "DF_subtiwiki.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_subti_enrich = ica_data.compute_annotation_enrichment(DF_subtiwiki,'value')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich['source'] = 'GO'\n",
    "# DF_pathway_enrich['source'] = 'KEGG pathways'\n",
    "# DF_module_enrich['source'] = 'KEGG modules'\n",
    "# DF_subti_enrich['source'] = 'SubtiWiki'\n",
    "\n",
    "DF_GO_enrich.rename({'gene_ontology':'annotation'},axis=1, inplace=True)\n",
    "# DF_pathway_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "# DF_module_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "# DF_subti_enrich.rename({'value':'annotation'},axis=1, inplace=True)\n",
    "\n",
    "DF_enrichments = pd.concat([DF_GO_enrich, DF_pathway_enrich, DF_module_enrich, DF_subti_enrich])\n",
    "DF_enrichments.to_csv(path.join(data_dir,'functional_enrichments.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for single gene iModulons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "1. Artificial knock-out of single genes\n",
    "1. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The find_single_gene_imodulons function identifies iModulons that are likely dominated by a single gene.\n",
    "\n",
    "The iModulons identified by ``find_single_gene_imodulons`` may contain more than one gene, since a threshold-agnostic method is used to identify these iModulons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg_imods = ica_data.find_single_gene_imodulons(save=True)\n",
    "len(sg_imods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,mod in enumerate(sg_imods):\n",
    "    ica_data.rename_imodulons({mod:'SG_'+str(i+1)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>regulator</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>regulon_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>n_regs</th>\n",
       "      <th>single_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SG_1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SG_5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     regulator  pvalue  qvalue  precision  recall  f1score  TP  regulon_size  \\\n",
       "SG_1       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_2       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_3       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_4       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "SG_5       NaN     NaN     NaN        NaN     NaN      NaN NaN           NaN   \n",
       "\n",
       "      imodulon_size  n_regs single_gene  \n",
       "SG_1            NaN     NaN        True  \n",
       "SG_2            NaN     NaN        True  \n",
       "SG_3            NaN     NaN        True  \n",
       "SG_4            NaN     NaN        True  \n",
       "SG_5            NaN     NaN        True  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table[ica_data.imodulon_table.single_gene == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>operon</th>\n",
       "      <th>regulator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BSU_24620</th>\n",
       "      <td>0.519729</td>\n",
       "      <td>tasA</td>\n",
       "      <td>AL009126.3</td>\n",
       "      <td>BSU24620</td>\n",
       "      <td>2553081</td>\n",
       "      <td>2553866</td>\n",
       "      <td>-</td>\n",
       "      <td>major biofilm matrix component</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>P54507</td>\n",
       "      <td>Op906</td>\n",
       "      <td>SigA,SinR,RemA,AbrB,LutR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_weight gene_name   accession old_locus_tag    start      end  \\\n",
       "BSU_24620     0.519729      tasA  AL009126.3      BSU24620  2553081  2553866   \n",
       "\n",
       "          strand                    gene_product               COG uniprot  \\\n",
       "BSU_24620      -  major biofilm matrix component  Function unknown  P54507   \n",
       "\n",
       "          operon                 regulator  \n",
       "BSU_24620  Op906  SigA,SinR,RemA,AbrB,LutR  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon('SG_1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save iModulon object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.util import explained_variance\n",
    "from pymodulon.io import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add iModulon sizes and explained variance\n",
    "for im in ica_data.imodulon_names:\n",
    "    ica_data.imodulon_table.loc[im,'imodulon_size'] = len(ica_data.view_imodulon(im))\n",
    "    ica_data.imodulon_table.loc[im,'explained_variance'] = explained_variance(ica_data,imodulons=im)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will save your iModulon table, your thresholds, and any other information stored in the ica_data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, path.join('..','data','interim','bsu_raw.json.gz'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer to view and edit your iModulon table in excel, save it as a CSV and reload the iModulon as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.imodulon_table.to_csv(path.join('..','data','interim','imodulon_table_raw.csv'))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
